### PR TITLE
enhance energy-output with seismic moment

### PR DESCRIPTION
**Pull request by Thomas-Ulrich**
- add the seismic moment time history to the new energy output (from which we could post-process the moment rate).
- improve terminal output of energy output
- remove deprecated fortran moment and moment rate output

On tpv13 Training, the new seismic moment differs by 1% and the moment rate output is also quite different because it is calculated from the interpolated dof on each side of the fault, and not from the slip at the Gauss points.
Using a larger mesh, the differences become insignificant.

Example of new terminal output:
```
Thu Apr 14 11:01:48, Info:  Waiting for last wave field.
Thu Apr 14 11:01:48, Info:  Writing wave field at time 0.
Thu Apr 14 11:01:48, Info:  Writing wave field at time 0. Done.
Thu Apr 14 11:01:48, Info:  Writing free surface at time 0.
Thu Apr 14 11:01:48, Info:  Writing free surface at time 0. Done.
Thu Apr 14 11:01:48, Info:  Writing energy output at time 0
Thu Apr 14 11:01:48, Info:  Writing energy output at time 0 Done.
Thu Apr 14 11:01:48, Info:  Writing faultoutput at time 0.
Thu Apr 14 11:01:48, Info:  Writing faultoutput at time 0. Done.
Thu Apr 14 11:01:49, Info:  #max-updates since sync:  0  @  0
Thu Apr 14 11:01:52, Info:  Writing energy output at time 1
Thu Apr 14 11:01:52, Info:  Elastic energy (total, % kinematic, % potential):  3.27231e+15  , 32.2478  , 67.7522
Thu Apr 14 11:01:52, Info:  Frictional work (total, % static, % radiated):  3.43709e+15  , 57.1268  , 42.8732
Thu Apr 14 11:01:52, Info:  Seismic moment (without plasticity): 3.20392e+18  Mw: 6.26712
Thu Apr 14 11:01:52, Info:  Plastic moment (value, equivalent Mw, % total moment): 9.8793e+16  , 5.25982  , 2.99127
Thu Apr 14 11:01:52, Info:  Writing energy output at time 1 Done.
Thu Apr 14 11:01:52, Info:  Writing faultoutput at time 1.
Thu Apr 14 11:01:52, Info:  Writing faultoutput at time 1. Done.
Thu Apr 14 11:01:52, Info:  0.277435 TFLOPS (rank 0: 138.642 GFLOPS, average over ranks: 138.718 GFLOPS)
Thu Apr 14 11:01:53, Info:  Writing energy output at time 2
Thu Apr 14 11:01:53, Info:  Elastic energy (total, % kinematic, % potential):  1.37704e+16  , 30.8691  , 69.1309
Thu Apr 14 11:01:53, Info:  Frictional work (total, % static, % radiated):  1.48918e+16  , 54.724  , 45.276
Thu Apr 14 11:01:53, Info:  Seismic moment (without plasticity): 1.5531e+19  Mw: 6.72413
Thu Apr 14 11:01:53, Info:  Plastic moment (value, equivalent Mw, % total moment): 7.54371e+17  , 5.84839  , 4.63219
Thu Apr 14 11:01:53, Info:  Writing energy output at time 2 Done.
Thu Apr 14 11:01:53, Info:  Writing faultoutput at time 2.
Thu Apr 14 11:01:53, Info:  Writing faultoutput at time 2. Done.
Thu Apr 14 11:01:53, Info:  0.393151 TFLOPS (rank 0: 196.421 GFLOPS, average over ranks: 196.575 GFLOPS)
Thu Apr 14 11:01:55, Info:  Writing energy output at time 3
Thu Apr 14 11:01:55, Info:  Elastic energy (total, % kinematic, % potential):  2.86422e+16  , 34.3936  , 65.6064
Thu Apr 14 11:01:55, Info:  Frictional work (total, % static, % radiated):  3.19665e+16  , 53.3625  , 46.6375
Thu Apr 14 11:01:55, Info:  Seismic moment (without plasticity): 3.73589e+19  Mw: 6.97826
Thu Apr 14 11:01:55, Info:  Plastic moment (value, equivalent Mw, % total moment): 3.26579e+18  , 6.27266  , 8.03893
Thu Apr 14 11:01:55, Info:  Writing energy output at time 3 Done.
Thu Apr 14 11:01:55, Info:  Writing faultoutput at time 3.
```

### CODE REVIEW

**Comment by krenzland** [comment]
>  (also the code is not really nicely formated: @krenzland if you would apply clang-format on energyOutput.* I would be able to format nicely also my code).

It is already clang-format formatted :)

**Comment by Thomas-Ulrich** [comment]
Here is a difference in the moment rate on the very coarse tpv13 Training example:
The differences are too large, I wonder if I'm not doing something wrong
image
```
import matplotlib.pylab as plt
import pandas
import numpy as np

df = pandas.read_csv("tpv13b_energy.csv")
time = df["time"]
seismic_moment = df["seismic_moment"]
moment_rate = np.gradient(seismic_moment, time[1])
plt.plot(time, moment_rate, label="energy output")

Enf = np.loadtxt("tpv13b-EnF_0t-all.dat")
plt.plot(Enf[:, 0], Enf[:, 1], label="fortran output")
plt.xlabel("time (s)")
plt.ylabel("moment rate (Nm/s)")
plt.legend()
plt.show()
```

**Comment by Thomas-Ulrich** [comment]
Using a 1.6mio cells mesh, I get a much better agreement, suggesting that the delay was due to a totally unresolved setup.
image

**Comment by Thomas-Ulrich** [comment]
Using a 1.6mio cells mesh, I get a much better agreement, suggesting that the delay was due to a totally unresolved setup.
image

**Comment by Thomas-Ulrich** [comment]
Even without terminal output, this quite clutters the log. I wonder if that impacts the performance.
```
Wed Apr 13 19:05:54, Info:  Writing energy output at time 9.61
Wed Apr 13 19:05:54, Info:  Writing energy output at time 9.61 Done.
Wed Apr 13 19:05:54, Info:  4.21268 TFLOPS (rank 0: 258.958 GFLOPS, average over ranks: 263.293 GFLOPS)
Wed Apr 13 19:05:54, Info:  Writing energy output at time 9.62
Wed Apr 13 19:05:55, Info:  Writing energy output at time 9.62 Done.
Wed Apr 13 19:05:55, Info:  4.21295 TFLOPS (rank 0: 258.974 GFLOPS, average over ranks: 263.309 GFLOPS)
Wed Apr 13 19:05:55, Info:  Writing energy output at time 9.63
Wed Apr 13 19:05:55, Info:  Writing energy output at time 9.63 Done.
Wed Apr 13 19:05:55, Info:  4.21321 TFLOPS (rank 0: 258.99 GFLOPS, average over ranks: 263.326 GFLOPS)
Wed Apr 13 19:05:56, Info:  Writing energy output at time 9.64
Wed Apr 13 19:05:56, Info:  Writing energy output at time 9.64 Done.
Wed Apr 13 19:05:56, Info:  4.21348 TFLOPS (rank 0: 259.006 GFLOPS, average over ranks: 263.342 GFLOPS)
Wed Apr 13 19:05:56, Info:  Writing energy output at time 9.65
Wed Apr 13 19:05:56, Info:  Writing energy output at time 9.65 Done.
```

**Comment by krenzland** [comment]
They now work like every other writer works. For my use-case, a relatively coarse sampling rate (similar to free surface output) is good enough. In this case, the cost doesn*t really matter because sync-points are pretty expensive anyway.  But without a sync point we can't/shouldn't do a reduce.
Does the Fortran output update more often?

**Comment by Thomas-Ulrich** [comment]
The current Fortran output is badly written.
An ASCII file written  for each rank, and the moment rate get written (without buffer) once every n dt with dt varying across ranks.
Then we use https://github.com/SeisSol/SeisSol/blob/master/postprocessing/science/concatenate_EnF_t.py to concatenate...
@ravil-mobile ported the fortran moment rate output to dr/cpp in https://github.com/SeisSol/SeisSol/pull/513 (https://github.com/SeisSol/SeisSol/pull/513/commits/a97b3ffa9c45c4ddd23147837fe15841acacf137) not yet merged.
At the end, I doubt this new sync-based would really impact performance. 
0.01s is sufficient, and it is typically what we use for surface output.

**Comment by krenzland** [comment]
If the surface output doesn't inhibit performance, the energy output won't either. (Well, that's a guess. But the computation isn't that expensive - 1 mesh traversal, some integrals, one reduce)

**Comment by krenzland** [comment]
It's definitely faster than writing one file per rank - that can be extremely bad on large runs. (Same actually true for receiver output as well)

**Comment by Thomas-Ulrich** [comment]
So I removed the old moment and moment rate output, and improved the terminal output:
```
Thu Apr 14 11:01:48, Info:  Waiting for last wave field.
Thu Apr 14 11:01:48, Info:  Writing wave field at time 0.
Thu Apr 14 11:01:48, Info:  Writing wave field at time 0. Done.
Thu Apr 14 11:01:48, Info:  Writing free surface at time 0.
Thu Apr 14 11:01:48, Info:  Writing free surface at time 0. Done.
Thu Apr 14 11:01:48, Info:  Writing energy output at time 0
Thu Apr 14 11:01:48, Info:  Writing energy output at time 0 Done.
Thu Apr 14 11:01:48, Info:  Writing faultoutput at time 0.
Thu Apr 14 11:01:48, Info:  Writing faultoutput at time 0. Done.
Thu Apr 14 11:01:49, Info:  #max-updates since sync:  0  @  0
Thu Apr 14 11:01:52, Info:  Writing energy output at time 1
Thu Apr 14 11:01:52, Info:  Elastic energy (total, % kinematic, % potential):  3.27231e+15  , 32.2478  , 67.7522
Thu Apr 14 11:01:52, Info:  Frictional work (total, % static, % radiated):  3.43709e+15  , 57.1268  , 42.8732
Thu Apr 14 11:01:52, Info:  Seismic moment (without plasticity): 3.20392e+18  Mw: 6.26712
Thu Apr 14 11:01:52, Info:  Plastic moment (value, equivalent Mw, % total moment): 9.8793e+16  , 5.25982  , 2.99127
Thu Apr 14 11:01:52, Info:  Writing energy output at time 1 Done.
Thu Apr 14 11:01:52, Info:  Writing faultoutput at time 1.
Thu Apr 14 11:01:52, Info:  Writing faultoutput at time 1. Done.
Thu Apr 14 11:01:52, Info:  0.277435 TFLOPS (rank 0: 138.642 GFLOPS, average over ranks: 138.718 GFLOPS)
Thu Apr 14 11:01:53, Info:  Writing energy output at time 2
Thu Apr 14 11:01:53, Info:  Elastic energy (total, % kinematic, % potential):  1.37704e+16  , 30.8691  , 69.1309
Thu Apr 14 11:01:53, Info:  Frictional work (total, % static, % radiated):  1.48918e+16  , 54.724  , 45.276
Thu Apr 14 11:01:53, Info:  Seismic moment (without plasticity): 1.5531e+19  Mw: 6.72413
Thu Apr 14 11:01:53, Info:  Plastic moment (value, equivalent Mw, % total moment): 7.54371e+17  , 5.84839  , 4.63219
Thu Apr 14 11:01:53, Info:  Writing energy output at time 2 Done.
Thu Apr 14 11:01:53, Info:  Writing faultoutput at time 2.
Thu Apr 14 11:01:53, Info:  Writing faultoutput at time 2. Done.
Thu Apr 14 11:01:53, Info:  0.393151 TFLOPS (rank 0: 196.421 GFLOPS, average over ranks: 196.575 GFLOPS)
Thu Apr 14 11:01:55, Info:  Writing energy output at time 3
Thu Apr 14 11:01:55, Info:  Elastic energy (total, % kinematic, % potential):  2.86422e+16  , 34.3936  , 65.6064
Thu Apr 14 11:01:55, Info:  Frictional work (total, % static, % radiated):  3.19665e+16  , 53.3625  , 46.6375
Thu Apr 14 11:01:55, Info:  Seismic moment (without plasticity): 3.73589e+19  Mw: 6.97826
Thu Apr 14 11:01:55, Info:  Plastic moment (value, equivalent Mw, % total moment): 3.26579e+18  , 6.27266  , 8.03893
Thu Apr 14 11:01:55, Info:  Writing energy output at time 3 Done.
Thu Apr 14 11:01:55, Info:  Writing faultoutput at time 3.
Thu Apr 14 11:01:55, Info:  Writing faultoutput at time 3. Done.

```

**Comment by Thomas-Ulrich** [comment]
updated the top message with all changes.

File: src/Kernels/DynamicRupture.cpp [hunk 1]
```diff
@@ -308,6 +318,11 @@ void seissol::kernels::DynamicRupture::flopsGodunovState( DRFaceInformation cons
   o_nonZeroFlops += dynamicRupture::kernel::computeTractionInterpolated::NonZeroFlops;
   o_hardwareFlops += dynamicRupture::kernel::computeTractionInterpolated::HardwareFlops;
 
+  o_nonZeroFlops += dynamicRupture::kernel::computeSquaredNormSlipRateInterpolated::NonZeroFlops;
+  o_hardwareFlops += dynamicRupture::kernel::computeSquaredNormSlipRateInterpolated::HardwareFlops;
+  o_nonZeroFlops += 2*tensor::squaredNormSlipRateInterpolated::size();
```
Comment (Thomas-Ulrich) [inline src/Kernels/DynamicRupture.cpp:323]: 2 or 3? + * and sqrt?

Comment (krenzland) [reply src/Kernels/DynamicRupture.cpp:323]: Good question. @sebwolf-de , @ravil-mobile what do you do? We oculd count sqrt as a single floating point operation. This undercounts, but defining how many flops a sqrt takes is compilcated (and platform dependent)

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks good, not sure how we should count the flops though...

It's a good idea to print the energies only when they are not *exactly* zero but this could break very easily due to floating point inaccuracies. Not sure what we should do here. Maybe check against 100*float epsilon or something similar?

**Comment by Thomas-Ulrich** [comment]
I think there is no floating point inaccuracy on a sum of zeros.

File: src/Kernels/DynamicRupture.cpp (hunk 1 above)
Comment (Thomas-Ulrich) [reply src/Kernels/DynamicRupture.cpp:323]: just found that on the web:
>A square root operation can take about 6 flops on typical CPU architectures, but [following [TBI97]], we will treat it as a single flop.
https://quick-adviser.com/how-do-you-calculate-number-of-flops/

Comment (sebwolf-de) [reply src/Kernels/DynamicRupture.cpp:323]: We do not count flops for the output (yet), the flops for the fortran output have not been counted either.
I'd count sqrt as 1 FLOP.

**Pull request merged**